### ISSUE TITLE
bug: Throat causes a halt in execution under certain conditions 

### ISSUE THREAD

**Issue by bjornua** [ForbesLindesay/throat#61]
**Describe the bug**

Throat causes a halt in execution under certain conditions

**To Reproduce**

```javascript
const _ = require('lodash');
const throat = require('throat');

async function main() {
  const worker = throat(4, async function (page) {});
  const pagesChunks = _.range(10).map((n) => _.range(100));

  let awaitCount = 0;
  for (const pagesChunk of pagesChunks) {
    console.log(`Await ${awaitCount++}`);
    await Promise.all(pagesChunk.map(worker));
  }

  console.log('Done');
}

main();
```

This outputs:
```text
Await 0
Await 1
Await 2
```

**Expected behaviour**

Expected output:

```text
Await 0
Await 1
Await 2
Await 3
Await 4
Await 5
Await 6
Await 7
Await 8
Await 9
Done
```

**Versions**

- node version: v14.17.5
- npm/yarn version: 6.14.14
- package version: throat-6.0.1

**Additional context**

Messing around with the length of `pagesChunk` or `size` can make the script complete normally, or halt earlier or later in the iterations.

**Comment by andriimaryskevych** [comment]
We faced the same issue with this library. @ForbesLindesay please take a look at this - the library has a critical bug and it has 18m+ weekly downloads

**Comment by peterlundberg** [comment]
Can confirm and isolate this a bit more also without the chunking above. 

```javascript
async function demoExitBug() {
  const limitedConcurrency = throat(1);
  for (let i = 0; i < 100; i += 1) {
    console.log('working on', i);
    await Promise.all([
      limitedConcurrency(async () => true), // with just one it would complete
      limitedConcurrency(async () => true), // concurrency + 1 = only 64 ok
      // limitedConcurrency(async () => true), // concurrency + 2 = only 32 ok
    ]);
  }
  console.log('Done!'); // This will not be reached in 6.0.1
}
demoExitBug();
```

**Comment by ForbesLindesay** [comment]
Sorry, didn't notice the issue. Pinging me on twitter is often good for super high priority issues like this. I'll push out a fix shortly.

**Comment by ForbesLindesay** [comment]
This is resolved in v6.0.2
